[quant][fix] Fix bug for ave pooling in FX quant

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73054

Summary:
Fix bug for ave pooling in FX quant

For the operator like avg pool, when user used prepare_custom_config_dict={"input_quantized_idxs": [0]}), then we don't insert observer for the input node. But the func 'maybe_make_input_output_share_observer' would check if input and output node need to share observer or not. if they are not shared successully, the output observer has to be removed. But for avg pool with the custimized config, we should keep the output observer.

solution:
Check if the observer need to be removed or not ( based on the condition ( output node + placeholder) , then check maybe_make_input_output_share_observers true / false.

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

terrychenism committed:
[quant][fix] Fix bug for ave pooling in FX quant

Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1169,6 +1169,8 @@ def insert_observers_for_model(
                             # to make all inputs and outputs use the first input's
                             # observer
                             if is_general_tensor_value_op or is_general_tensor_shape_op or is_reuse_input_qconfig_:
+                                if get_arg_target_dtype_as_output(node, modules, node_name_to_target_dtype) == torch.quint8:
```
can we merge this with the condition in L1171?

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34334059).

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -162,6 +162,13 @@ def is_output_dtype_supported_by_backend(
     return output_dtype is None or \
         output_dtype == node_name_to_target_dtype[node.name]["output_activation_dtype"]
 
+def is_need_remove_observer(node, modules, node_name_to_target_dtype):
```
nit: `is_need_remove_observer` --> `need_remove_observer`

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5006,6 +5006,47 @@ def forward(self, x):
+        count_check = {
+            ns.call_method('dequantize') : 1
+        }
+        order_check = [
+            ns.call_method('dequantize'),
+        ]
```
should we add more ops? like avgpool -> dequantize

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34334059).

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -162,6 +162,13 @@ def is_output_dtype_supported_by_backend(
     return output_dtype is None or \
         output_dtype == node_name_to_target_dtype[node.name]["output_activation_dtype"]
 
+def need_remove_observer(node, modules, node_name_to_target_dtype):
```
can we add some comments on what this is doing and why?

vkuzo replied:
would it be possible to name this function after what it is checking?  `need_remove_observer` is pretty generic and it may not be clear on why exactly this function is there.

I think a PR summary explaining the issue and the fix might help me give better comments here, right now I'm not yet sure what exactly is broken.

vkuzo commented:
Could we add some more context to the PR summary on what the bug is, and what this diff is doing conceptually to fix it?

vkuzo commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5006,6 +5006,48 @@ def forward(self, x):
+                x = self.avg_pool2d(x)
+                return x
+
+        # This model is not executable since we just put all ops
```
should we make these comments reflect the code?  Seems like these are copy-pasted, but not sure if they reflect this callsite.

jerryzh168 replied:
this comment is copied I think, we can remove it

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34334059).

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34334059).

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34334059).

jerryzh168 commented:
please add more context in the summary that describes what the bug is, and how we fixed it

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -162,6 +162,16 @@ def is_output_dtype_supported_by_backend(
         output_dtype == node_name_to_target_dtype[node.name]["output_activation_dtype"]
 
+def need_remove_observer(node, modules, node_name_to_target_dtype):
+    """ remove output observer when the node output is quint8
```
I think we can give more context on why here

jerryzh168 replied:
we first need to explain when do we remove observers, we remove it when `maybe_make_input_output_share_observers` returns False, which means the function fails to share output observer and input observer, and we can assume that input is not observed and remove the output observer.

When node output is not fp32 and it is placeholder, the input is assumed to be quantized, so it is observed in a different place, rather than not observed, therefore, even `maybe_make_input_output_share_observers` returns False, we still can't remove the observer.

This is a bit complicated, we can discuss offline if needed.

terrychenism replied:
i think first we need to check the observer need to be removed or not ( based on the condition ( output node + placeholder) , then check  maybe_make_input_output_share_observers true / false.

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1165,10 +1175,13 @@ def insert_observers_for_model(
-                            if is_general_tensor_value_op or is_general_tensor_shape_op or is_reuse_input_qconfig_:
+                            if (is_general_tensor_value_op and need_remove_observer_) or \
+                                    is_general_tensor_shape_op or is_reuse_input_qconfig_:
                                 if not maybe_make_input_output_share_observers(node, model, modules):
```
I feel it makes more sense to put `need_remove_observer_` here?

terrychenism replied:
we cannot put it here, cuz it would make conflicts on some cases with the flag  'is_reuse_input_qconfig_'

jerryzh168 replied:
I see, this is getting pretty complicated, feels like we need some more refactors and docs..

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34334059).

vkuzo reviewed (commented):
Could we add a description to the PR summary of what was actually broken, and what this PR is conceptually doing to fix it?

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

For the operator like avg pool, when user used prepare_custom_config_dict={"input_quantized_idxs": [0]}), then we don't insert observer for the input node. But the func 'maybe_make_input_output_share_observer' would check if input and output node need to share observer or not. if they are not shared successully, the output observer has to be removed. But for avg pool with the custimized config, we should keep the output observer.

solution:
Check if the observer need to be removed or not ( based on the condition ( output node + placeholder) , then check maybe_make_input_output_share_observers true / false.

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fix] Fix bug for ave pooling in FX quant"


Summary:
Fix bug for ave pooling in FX quant

For the operator like avg pool, when user used prepare_custom_config_dict={"input_quantized_idxs": [0]}), then we don't insert observer for the input node. But the func 'maybe_make_input_output_share_observer' would check if input and output node need to share observer or not. if they are not shared successully, the output observer has to be removed. But for avg pool with the custimized config, we should keep the output observer.

solution:
Check if the observer need to be removed or not ( based on the condition ( output node + placeholder) , then check maybe_make_input_output_share_observers true / false.

Test Plan:
python3 test/test_quantization.py TestQuantizeFxOps.test_ave_pool_with_custom_cfg

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34334059](https://our.internmc.facebook.com/intern/diff/D34334059)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34334059).
